File: tests/brain/python/circuit.py
# Copyright (c) 2016, EPFL/Blue Brain Project
#                     Juan Hernando <juan.hernando@epfl.ch>
#
# This file is part of Brion <https://github.com/BlueBrain/Brion>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import setup
import brain
import numpy

import unittest

class TestCircuitOpen(unittest.TestCase):

    def test_bad_open(self):
        self.assertRaises(RuntimeError, lambda: brain.Circuit("foo"))

    def test_open(self):
        circuit = brain.Circuit(brain.test.circuit_config)

class TestCircuit(unittest.TestCase):
    def setUp(self):
        self.circuit = brain.Circuit(brain.test.circuit_config)

    def test_gids(self):
        assert(self.circuit.num_neurons() == 1000)
        gids = self.circuit.gids('Column')
        assert(len(gids) == 1000)
        for gid, index in zip(gids, range(1000)):
            assert(gid == index + 1)

        assert(len(self.circuit.gids()) == 1000)

        assert(len(self.circuit.random_gids(0.1)) == 100)
        assert(len(self.circuit.random_gids(0.1, 'Column')) == 100)

    def test_iteration_order(self):

        def get(functor, *args):
            a = []
            a += list(functor(self.circuit, [100], *args))
            a += list(functor(self.circuit, [20, 40], *args))
            a += list(functor(self.circuit, [7], *args))
            a += list(functor(self.circuit, [87, 89, 88], *args))
            a += list(functor(self.circuit, {997, 123, 1}, *args))

            b = list(functor(self.circuit, [100, 20, 40, 7, 87, 89, 88] +
                             list({997, 123, 1}), *args))
            return a, b

        a, b = get(brain.Circuit.morphology_types)
        assert(a == b)
        a, b = get(brain.Circuit.morphology_uris)
        assert(a == b)
        a, b = get(brain.Circuit.load_morphologies,
                   brain.Circuit.Coordinates.local)
        assert(len(a) == len(b))
        for i, j in zip(a, b):
            assert(numpy.all(i.points() == j.points()))

    def test_morphology_uris(self):
        uris = self.circuit.morphology_uris([1, 100, 1000])
        assert['sm090227a1-2_idC.h5' in uris[0]]
        assert['sm101103b1-2_INT_idC.h5' in uris[1]]
        assert['tkb060509a2_ch3_mc_n_el_100x_1.h5' in uris[2]]

    def test_geometry(self):
        gids = self.circuit.gids()
        transforms = self.circuit.transforms(gids)
        positions = self.circuit.positions(gids)
        rotations = self.circuit.rotations(gids)
        assert(transforms.shape == (1000, 4, 4))
        assert(positions.shape == (1000, 3))
        assert(rotations.shape == (1000, 4))

    def test_load_morphology(self):
        morphologies = self.circuit.load_morphologies(
            [1, 100, 1000], brain.Circuit.Coordinates.local)
        morphologies = self.circuit.load_morphologies(
            [1, 100, 1000], brain.Circuit.Coordinates.global_)

    def test_metypes(self):
        count = self.circuit.num_neurons()
        mnames = self.circuit.morphology_type_names()
        assert(len(mnames) == 9)
        for t in ['L1_SLAC', 'L23_PC', 'L23_MC', 'L4_PC', 'L4_MC',
                  'L5_TTPC1', 'L5_MC', 'L6_TPC_L1', 'L6_MC']:
            assert(t in mnames)

        gids = self.circuit.gids('L23_PC')
        mtypes = self.circuit.morphology_types(gids)
        assert(len(mtypes) == len(gids))
        for t in mtypes:
            assert(mnames[t] == 'L23_PC')

        enames = self.circuit.electrophysiology_type_names()
        assert(len(enames) == 2)
        for t in ['cACint', 'cADpyr']:
            assert(t in enames)
        etypes = self.circuit.electrophysiology_types(gids)
        assert(len(etypes) == len(gids))
        for t in etypes:
            assert(enames[t] == 'cADpyr')

    def test_repeated_gid(self):
        self.assertRaises(ValueError,
                          lambda: self.circuit.positions([1, 3, 1]))

    def test_bad_gids(self):
        self.assertRaises(RuntimeError,
                          lambda: self.circuit.positions(numpy.array([0])))
        self.assertRaises(RuntimeError,
                          lambda: self.circuit.morphology_uris(
                              numpy.array([10000])))
        self.assertRaises(RuntimeError, lambda: self.circuit.positions([0]))
        self.assertRaises(RuntimeError,
                          lambda: self.circuit.load_morphologies(
                              [100000], brain.Circuit.Coordinates.local))

if __name__ == '__main__':
    unittest.main()

File: brain/python/circuit.cpp
/* Copyright (c) 2013-2016, EPFL/Blue Brain Project
 *                          Juan Hernando <juan.hernando@epfl.ch>
 *
 * This file is part of Brion <https://github.com/BlueBrain/Brion>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <boost/python.hpp>
#include <boost/python/stl_iterator.hpp>

#include "arrayHelpers.h"
#include "docstrings.h"
#include "helpers.h"
#include "types.h"

#include <brain/circuit.h>
#include <brain/synapses.h>
#include <brain/synapsesStream.h>

namespace bp = boost::python;

namespace brain
{
namespace
{
/* Reorders a collection with a random access iterator in place provided a pair
   of iterators with the desired order. The order collection is overwritten.
   http://stackoverflow.com/questions/838384
*/
template <typename T>
void _reorderDestructive(std::vector<T>& values, uint32_ts& order)
{
    size_t remaining = order.size();
    for (uint32_t in_pos = 0; remaining > 0; ++in_pos)
    {
        uint32_t out_pos = order[in_pos];
        if (out_pos == std::numeric_limits<uint32_t>::max())
            continue;
        --remaining;
        T tmp = values[in_pos];
        for (uint32_t i; out_pos != in_pos; out_pos = i, --remaining)
        {
            assert(out_pos != std::numeric_limits<uint32_t>::max());
            std::swap(tmp, values[out_pos]);
            std::swap(order[out_pos], i = std::numeric_limits<uint32_t>::max());
        }
        values[in_pos] = tmp;
    }
}

template <typename MemberFunction>
bp::object _getProperty(const Circuit& circuit, const MemberFunction& property,
                        bp::object cellSet)
{
    uint32_ts mapping;
    GIDSet gids;
    gidsFromPython(cellSet, gids, mapping);
    auto values = (circuit.*property)(gids);
    if (!mapping.empty())
        _reorderDestructive(values, mapping);
    return toNumpy(std::move(values));
}

CircuitPtr Circuit_initFromURI(const std::string& uri)
{
    return CircuitPtr(new Circuit(URI(uri)));
}

bp::object Circuit_getAllGIDs(const Circuit& circuit)
{
    return toNumpy(toVector(circuit.getGIDs()));
}

bp::object Circuit_getGIDs(const Circuit& circuit, const std::string& target)
{
    return toNumpy(toVector(circuit.getGIDs(target)));
}

bp::object Circuit_getRandomTargetGIDs(const Circuit& circuit,
                                       const float fraction,
                                       const std::string& target)
{
    return toNumpy(toVector(circuit.getRandomGIDs(fraction, target)));
}

bp::object Circuit_getRandomGIDs(const Circuit& circuit, const float fraction)
{
    return toNumpy(toVector(circuit.getRandomGIDs(fraction)));
}

#define GET_CIRCUIT_PROPERTY_FOR_GIDS(property)                               \
    bp::object Circuit_get##property(const Circuit& circuit, bp::object gids) \
    {                                                                         \
        return _getProperty(circuit, &Circuit::get##property, gids);          \
    }

GET_CIRCUIT_PROPERTY_FOR_GIDS(MorphologyTypes)
GET_CIRCUIT_PROPERTY_FOR_GIDS(ElectrophysiologyTypes)

bp::object Circuit_getMorphologyURIs(const Circuit& circuit, bp::object object)
{
    // Not trying to use a more general version of _getProperty because of its
    // complexity.
    uint32_ts mapping;
    GIDSet gids;
    gidsFromPython(object, gids, mapping);
    auto uris = circuit.getMorphologyURIs(gids);
    if (!mapping.empty())
        _reorderDestructive(uris, mapping);
    return toPythonList(uris);
}

#define GET_CIRCUIT_PROPERTY_VALUES(property)              \
    bp::list Circuit_get##property(const Circuit& circuit) \
    {                                                      \
        return toPythonList(circuit.get##property());      \
    }
GET_CIRCUIT_PROPERTY_VALUES(MorphologyTypeNames)
GET_CIRCUIT_PROPERTY_VALUES(ElectrophysiologyTypeNames)

bp::list Circuit_loadMorphologies(const Circuit& circuit, bp::object object,
                                  Circuit::Coordinates coords)
{
    // Not trying to use a more general version of _getProperty because of its
    // complexity.
    uint32_ts mapping;
    GIDSet gids;
    gidsFromPython(object, gids, mapping);
    auto morphologies = circuit.loadMorphologies(gids, coords);
    if (!mapping.empty())
        _reorderDestructive(morphologies, mapping);
    return toPythonList(morphologies);
}

bp::object Circuit_getPositions(const Circuit& circuit, bp::object gids)
{
    return _getProperty(circuit, &Circuit::getPositions, gids);
}

bp::object Circuit_getTransforms(const Circuit& circuit, bp::object gids)
{
    bp::object matrices = _getProperty(circuit, &Circuit::getTransforms, gids);
    // We want the result to be indexed using regular mathematical notation
    // even if the actual storage is column-major.
    return matrices.attr("transpose")(0, 2, 1);
}

bp::object Circuit_getRotations(const Circuit& circuit, bp::object gids)
{
    return _getProperty(circuit, &Circuit::getRotations, gids);
}

SynapsesWrapper Circuit_getAfferentSynapses(
    const CircuitPtr& circuit, bp::object gids,
    const brain::SynapsePrefetch prefetch)
{
    return SynapsesWrapper(circuit->getAfferentSynapses(gidsFromPython(gids),
                                                        prefetch),
                           circuit);
}

SynapsesWrapper Circuit_getExternalAfferentSynapses(
    const CircuitPtr& circuit, bp::object gids, const std::string& source,
    const brain::SynapsePrefetch prefetch)
{
    return SynapsesWrapper(circuit->getExternalAfferentSynapses(
                               gidsFromPython(gids), source, prefetch),
                           circuit);
}

SynapsesWrapper Circuit_getEfferentSynapses(
    const CircuitPtr& circuit, bp::object gids,
    const brain::SynapsePrefetch prefetch)
{
    return SynapsesWrapper(circuit->getEfferentSynapses(gidsFromPython(gids),
                                                        prefetch),
                           circuit);
}

SynapsesWrapper Circuit_getProjectedSynapses(
    const CircuitPtr& circuit, bp::object pre, bp::object post,
    const brain::SynapsePrefetch prefetch)
{
    return SynapsesWrapper(circuit->getProjectedSynapses(gidsFromPython(pre),
                                                         gidsFromPython(post),
                                                         prefetch),
                           circuit);
}
}

void export_Circuit()
// clang-format off
{

bp::class_<Circuit, boost::noncopyable, CircuitPtr> circuitWrapper(
    "Circuit", DOXY_CLASS(brain::Circuit), bp::no_init);

bp::scope circuitScope = circuitWrapper;

bp::enum_<Circuit::Coordinates>("Coordinates")
    .value("global_", Circuit::Coordinates::global)
    .value("local", Circuit::Coordinates::local);

const auto selfarg = bp::arg("self");

// Do not modify whitespace on DOXY_FN lines
circuitWrapper
    .def("__init__", bp::make_constructor(Circuit_initFromURI),
         DOXY_FN(brain::Circuit::Circuit(const URI&)))
    .def("gids", Circuit_getAllGIDs, (selfarg),
         DOXY_FN(brain::Circuit::getGIDs() const))
    .def("gids", Circuit_getGIDs, (selfarg, bp::arg("target")),
         DOXY_FN(brain::Circuit::getGIDs(const std::string&) const))
    .def("random_gids", Circuit_getRandomTargetGIDs,
         (selfarg, bp::arg("fraction"), bp::arg("target")),
         DOXY_FN(brain::Circuit::getRandomGIDs(float, const std::string&) const))
    .def("random_gids", Circuit_getRandomGIDs,
         (selfarg, bp::arg("fraction")),
         DOXY_FN(brain::Circuit::getRandomGIDs(float) const))
    .def("morphology_uris", Circuit_getMorphologyURIs,
         (selfarg, bp::arg("gids")),
         DOXY_FN(brain::Circuit::getMorphologyURIs))
    .def("load_morphologies", Circuit_loadMorphologies,
         (selfarg, bp::arg("gids"), bp::arg("coords")),
         DOXY_FN(brain::Circuit::loadMorphologies))
    .def("positions", Circuit_getPositions, (selfarg, bp::arg("gids")),
         DOXY_FN(brain::Circuit::getPositions))
    .def("morphology_types", Circuit_getMorphologyTypes,
         (selfarg, bp::arg("gids")),
         DOXY_FN(brain::Circuit::getMorphologyTypes))
    .def("morphology_type_names", Circuit_getMorphologyTypeNames, (selfarg),
         DOXY_FN(brain::Circuit::getMorphologyTypeNames))
    .def("electrophysiology_types", Circuit_getElectrophysiologyTypes,
         (selfarg, bp::arg("gids")),
         DOXY_FN(brain::Circuit::getElectrophysiologyTypes))
    .def("electrophysiology_type_names",
         Circuit_getElectrophysiologyTypeNames, (selfarg),
         DOXY_FN(brain::Circuit::getElectrophysiologyTypeNames))
    .def("transforms", Circuit_getTransforms, (selfarg, bp::arg("gids")),
         DOXY_FN(brain::Circuit::getTransforms))
    .def("rotations", Circuit_getRotations, (selfarg, bp::arg("gids")),
         DOXY_FN(brain::Circuit::getRotations))
    .def("num_neurons", &Circuit::getNumNeurons, (selfarg),
         DOXY_FN(brain::Circuit::getNumNeurons))
    .def("afferent_synapses", Circuit_getAfferentSynapses,
         (selfarg, bp::arg("gids"),
          bp::arg("prefetch") = SynapsePrefetch::none),
         DOXY_FN(brain::Circuit::getAfferentSynapses))
    .def("external_afferent_synapses", Circuit_getExternalAfferentSynapses,
         (selfarg, bp::arg("gids"), bp::arg("source"),
          bp::arg("prefetch") = SynapsePrefetch::none),
         DOXY_FN(brain::Circuit::getExternalAfferentSynapses))
    .def("efferent_synapses", Circuit_getEfferentSynapses,
         (selfarg, bp::arg("gids"),
          bp::arg("prefetch") = SynapsePrefetch::none),
         DOXY_FN(brain::Circuit::getEfferentSynapses))
    .def("projected_synapses", Circuit_getProjectedSynapses,
         (selfarg, bp::arg("preGids"), bp::arg("postGIDs"),
          bp::arg("prefetch") = SynapsePrefetch::none),
         DOXY_FN(brain::Circuit::getProjectedSynapses));
}
// clang-format on
}
